Terminology Resolver

SoerenHoffstedt opened the pull request:
[T531](https://dbs-projects.mathematik.uni-marburg.de/T531) TerminologyResolver Operator and Terminology class to use terminologies service from gfbio.

michaelmattig commented on src/operators/terminology_resolver.cpp:
```diff
@@ -0,0 +1,180 @@
+#include "operators/operator.h"
+
+#include <json/json.h>
+#include "../util/terminology.h"
```
include from base like the rest

michaelmattig commented on src/operators/terminology_resolver.cpp:
```diff
@@ -0,0 +1,180 @@
+ * Operator for resolving strings with the terminology service from gfbio: https://terminologies.gfbio.org/
+ *
+ * parameters:
+ *  - column: name of the textual column to resolve
```
@jdroenner @ChristianBeilschmidt would `attribute_name` be better here? We always talk about attributes not columns, like e.g. `textual_attribute_filter` and not `textual_column_filter`

jdroenner replied:
yes, attribute appears to be a better name.

michaelmattig commented on src/operators/terminology_resolver.cpp:
```diff
@@ -0,0 +1,180 @@
+ *         if requested field is an array, the first element will be returned.
+ *  - match_type: "exact", "included", "regex", see TerminologyService search API. "exact" if not provided.
+ *  - first_hit: bool, see TerminologyService search API. true if not provided.
+ *  - name_appendix: name of the next column for the resolved text values
```
> name of the *next* column for the resolved text values
should say *new* column I guess.

why is it called *appendix* if it can also replace the entire string? With this scheme it's also not possible to name the new column "teminology" ;) I'd propose calling the option `resolved_column` (or `resolved_attribute` depending on the issue above) and just accepting a custom string. the user can put the terminology name there if he wants to.

jdroenner replied:
we can also handle the appendix part in the UI. like everything...

michaelmattig commented on src/util/terminology.cpp:
```diff
@@ -0,0 +1,246 @@
+    typedef boost::shared_ptr<task_t> ptask_t;
+
+    //init thread pool
+    int threads_num = Configuration::get<int>("terminology.threads",16);
```
add to conf/settings-default.toml

michaelmattig commented on src/util/terminology.h:
```diff
@@ -0,0 +1,77 @@
+#include <Poco/Net/HTTPResponse.h>
+#include "datatypes/simplefeaturecollection.h"
+
+enum HandleNotResolvable {
```
rather use enum class

michaelmattig commented on src/util/terminology.cpp:
```diff
@@ -0,0 +1,246 @@
+                                            const bool first_hit,
+                                            const HandleNotResolvable on_not_resolvable) {
+
+    Poco::URI uri("https://terminologies.gfbio.org/api/terminologies/search");
```
load url from config to be able to replace e.g. with dev version of terminology server

michaelmattig commented on src/util/terminology.cpp:
```diff
@@ -0,0 +1,246 @@
+    {
+        Json::Value results = response_json["results"];
+
+        Json::Value val = results[0].get(key, not_resolved);
```
what if there was no hit?

SoerenHoffstedt replied:
It returns `not_resolved` if `results[0]`is out of bounds. But I'll make this more explicit by checking if results is empty.

michaelmattig commented on src/util/terminology.cpp:
```diff
@@ -0,0 +1,246 @@
+
+        Json::Value val = results[0].get(key, not_resolved);
+        if(val.isArray())
+            return std::make_pair(name, val[0].asString());
```
what if `val` is empty?

michaelmattig commented on test/systemtests/data/terminology_test.csv:
```diff
@@ -0,0 +1,16 @@
+7.58582,53.10799,deer,26842
+8.39915,49.28734,wolf,67354
+9.93641,49.79279,duck,97074
+8.27835,50.26977,cheese,65510
```
add newline

jdroenner commented on src/operators/terminology_resolver.cpp:
```diff
@@ -0,0 +1,180 @@
+ *      - if not provided the name will be "column resolved" with column being the parameter above.
+ *  - on_not_resolvable: if no label for the string was found, what to insert into new column
+ *      - "EMPTY"
+ *      - "OLD_NAME"
```
or just call it KEEP

SoerenHoffstedt commented on src/util/terminology.cpp:
```diff
@@ -0,0 +1,246 @@
+    if(threads_num > names_in.size())
+        threads_num = names_in.size();
+
+    boost::asio::io_service io_service;
```
The thread pool was added around boost version 1.66 and Ubuntu 16.04 is using version 1.58, so it is not available for us.

SoerenHoffstedt commented on src/operators/terminology_resolver.cpp:
```diff
@@ -0,0 +1,180 @@
+        names_in.push_back(old_attribute_array.get(i));
+    }
+
+    std::vector<std::string> names_out;
```
The reason I did this, isn't valid anymore after another change. I will adjust it.

SoerenHoffstedt commented:
I pushed changes addressing the comments. 
I changed `column` to `attribute` and `OLD_NAME` to `KEEP` in the parameters.

michaelmattig commented on test/unittests/terminology.cpp:
```diff
@@ -0,0 +1,39 @@
+
+#include "../../../../mapping-gfbio/src/util/terminology.h"
```
how did you execute this test? Currently, only tests in mapping-core are executed. Maybe you can try to look into including module tests in the same way as the module src is added to the core.

SoerenHoffstedt replied:
Ah, I didn't add the comment, sorry. I was executing this class from core while testing, but copied it to gfbio. It is not executable at the moment but I wanted to keep the test somewhere.
I will look into adding the test.

michaelmattig commented on test/unittests/terminology.cpp:
```diff
@@ -0,0 +1,39 @@
+    std::string second = "honey bee";
+    std::string term1 = "NCBITAXON";
+    std::string term2 = "PESI";
+    EXPECT_EQ(Terminology::resolveSingle(first, term1, "label", EMPTY), "Prunus domestica");
```
has to be adapted to the changed method signature

SoerenHoffstedt commented:
Fixed unittest and added variables so that mapping-core can add the provided unittests and systemtests to its target. Functionality of the tests depends on [pull request #6](https://github.com/umr-dbs/mapping-core/pull/6) in mapping-core.

michaelmattig commented:
I'm getting the following error when trying to run the system test using mapping_manager:
![grafik](https://user-images.githubusercontent.com/10283892/38372062-3d9142b0-38ee-11e8-800c-e7a85dcdb714.png)
can you reproduce this?

unfortunately, the unittets also don't work for me:
```
[ 98%] Linking CXX executable ../../target/bin/mapping_unittests
../../target/lib/libmapping_core_unittests_lib.a(terminology.cpp.o): In function `__static_initialization_and_destruction_0(int, int)':
/mnt/c/Users/Michael/CLionProjects/mapping-gfbio/test/unittests/terminology.cpp:5: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
/mnt/c/Users/Michael/CLionProjects/mapping-gfbio/test/unittests/terminology.cpp:14: undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
collect2: error: ld returned 1 exit status
make[3]: *** [../target/bin/mapping_unittests] Error 1
make[2]: *** [test/CMakeFiles/mapping_unittests.dir/all] Error 2
make[1]: *** [test/CMakeFiles/mapping_unittests.dir/rule] Error 2
make: *** [mapping_unittests] Error 2
```
any ideas?

michaelmattig commented:
I think the problem with the SEGFAULT of the system test may be this:
`std::unique_ptr<OGRGeometry> defGeom(defaultGeomPtr);`
When the unique pointer is dropped it tries to free a nullpointer. On the other hand, if there is a default geometry it would here use the default deallocator, but it should use `OGRFeature::DestroyFeature`. So I think you would need to use a custom deallocator for the unique pointer. Do we really need the unique pointer, though?

SoerenHoffstedt commented:
Additionally `OGRGeometry *defaultGeomPtr;` is uninitialized, that might cause the SEGFAULT. Because the scope of the default geometry is the function the `unique_ptr` is not really needed. I was working on this part of the code at the moment for the OGRSource, should I already push a fix just for this?
